hil/ext/switches/juniper.py

```diff
@@ -0,0 +1,404 @@
+
+# Copyright 2013-2017 Massachusetts Open Cloud Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+"""A switch driver for the juniper QFX series.
+
+But in theory can work for any Juniper switch that supports NETCONF.
+Driver uses the python library PyEZ made available by Juniper.
+By default it sets NETCONF session over ssh, i.e. make RPC calls over ssh.
+For this to work, make sure that the Juniper switch is configured
+appropriately to recieve NETCONF connections. Its default port is 830.
+"""
+
+import re
+import logging
+import schema
+import json
+import sys
+import os
+
+from jnpr.junos import Device
+from jnpr.junos.exception import ConnectError
+from jnpr.junos.utils.config import Config
+from jnpr.junos.utils.config import ConfigLoadError
+
+
+from hil.model import db, Switch
+from hil.model import BigIntegerType
+from hil.migrations import paths
+from os.path import dirname, join
+from hil.migrations import paths
+from hil.ext.switches.junos.config_tables.ConfigTables import (
+        InterfaceConfigTable
+        )
+logger = logging.getLogger(__name__)
+paths[__name__] = join(dirname(__file__), 'migrations', 'juniper')
+
+
+class VlanAddError(Exception):
+    pass
+
+
+class ConfigCommitError(Exception):
+    pass
+
+
+class Juniper(Switch):
```

Comment:
I don't even know where to begin with this class. It is an unreadable, umaintainable mess. I seriously can't follow it. A few concrete suggestions:

* Get rid of the global variables (enable_interface, add_vlan, etc). There's almost never a good reason for a mutable global variable in all but the tiniest programs.

* Rather than constructing the commands as strings, then passing them around untill they're eventually used, encapsulate the functionality you need in methods/functions -- the nexus driver provides a number of good examples of what this looks like:

```python
    def disable_vlan(self, vlan_id):
        self.console.sendline('sw trunk allowed vlan remove %s' % vlan_id)
```

Then just call the functions where you need them.

* Anywhere where you've got something like:

```python
try:
    with ...:
        try:
            ...
...
```

...or similar is a bad sign; that much non-local control flow is almost guaranteed to be unreadable. Simplify that sort of thing.

* You've got some kind of session related to the library that you're passing around, but the session() method just returns self -- might it make sense to actually return something else, and store some of the state there? I really can't follow this, so it's hard for me to be more specific.

Reply:
Here is the context, after getting the initial driver working. I took a second look at it to see if I can remove repetitive chunks of the code. I realized that there were some common set of commands that I was using to do most of the port related changes in the switch. 
Also, the some commands were long and had to be on a single line but I also had to stick to the 79 character limit. 

Considering this factors I did three things.
1. defined bunch of variables that other functions can use and initialized them to be None.
Their names reflect the commands that they will contain:
```
    enable_interface = None
    add_vlan = None
    remove_vlan = None
    set_native_vlan = None
    remove_native_vlan = None
    remove_default_vlan = None
    set_trunk_mode = None
```
2. defined a method that will initialized them for a given port and vlan
```
    def _set_commands(self, port, vlan_id):
        """ Set of common `set` commands used by other functions."""
```
3. and the wrote another function to clear the state at the end of each use.

I initially was reluctant to use global variables but did not find a better way to implement this. I will look at the nexus driver and see if I can improve this.

Reply:
For the long-line problem, you can do something like this:

```python
"foo bar baz quux " \
    "apple bannana foo " \
    "more stuff"
```

Python will concatenate string literals that are next to one another.

Like I said, I think the way to go is to just have a bunch of methods, rather than string variables, which take parameters (and the string construction happens in the method).